inc/Config/QueryContext/HttpQueryContext.php

```diff
@@ -180,4 +242,49 @@ public function process_response( string $raw_response_data, array $input_variab
 	final public function is_response_data_collection(): bool {
 		return $this->output_schema['is_collection'] ?? false;
 	}
+
+	// @todo: consider splitting the datasource injection out from query context so we don't have to tie a query
+	// to a datasource when instantiating. instead, we can just require applying queries to datasources in query
+	// runner execution. ie: $query_runner->execute( $query, $datasource );
+	//
+	/** @psalm-suppress ParamNameMismatch reason: we want the clarity provided by the rename here */
+	final public static function from_array( array $config, ?ValidatorInterface $validator = null ): self|\WP_Error {
+		if ( ! isset( $config['datasource'] ) || ! $config['datasource'] instanceof HttpDatasourceInterface ) {
+			return new \WP_Error( 'missing_datasource', __( 'Missing datasource.', 'remote-data-blocks' ) );
+		}
+
+		$validator = $validator ?? new Validator( self::CONFIG_SCHEMA );
+		$validated = $validator->validate( $config );
+
+		if ( is_wp_error( $validated ) ) {
+			return $validated;
+		}
+
+		return new self( $config['datasource'], $config['input_schema'], $config['output_schema'], $config );
+	}
+
+	public function to_array(): array {
```

Comment:
What is the use case for this method? Why doesn't this just return the private $config? And should this be `final`?

Reply:
It's part of `ArraySerializableInterface`. We could split into serializable/unserializable if we find having both together is a poor assumption, but, currently it's that way because it's a fairly common arrangement and most abstractions can probably conjure some sort of output even if they don't use it much.

Sure, it could return `$config`. I just stuck things that would never be privileged in any way in there for now because there's no immediate use case and I had bigger fish to fry.